File: dice.py
import numpy as np

# Part A

# 1. Calculate Total Combinations
total_combinations = 6 * 6  # Total combinations when rolling two six-sided dice
print("Total combinations:", total_combinations)

# 2. Calculate Distribution of All Possible Combinations
distribution = np.zeros(13, dtype=int)
for i in range(1, 7):
    for j in range(1, 7):
        sum_val = i + j
        distribution[sum_val] += 1  # Increment count for each sum
print("Distribution of combinations:", distribution[2:])

# 3. Calculate Probability of All Possible Sums
probability = distribution[2:] / total_combinations  # Probability = count / total
print("Probability of sums:", probability)

# Part B

# Function to Reattach Spots on the Dice while maintaining probabilities
def undoom_dice(Die_A, Die_B):
    New_Die_A = []
    New_Die_B = []
    sum_counts = np.zeros(11, dtype=int)  # Array to store counts of sums 2 to 12

    # Loop through all combinations of Die A and Die B
    for i in range(len(Die_A)):
        for j in range(len(Die_B)):
            sum_val = Die_A[i] + Die_B[j]
            sum_counts[sum_val - 2] += 1  # Increment count for this sum (2 to 12)
            if Die_A[i] <= 4:  # Check condition for Die A
                New_Die_A.append(Die_A[i])  # Keep Die A as is if spots <= 4
                New_Die_B.append(Die_B[j])  # Keep Die B as is

    # Adjust Die B if necessary to maintain probabilities
    for i in range(len(New_Die_B)):
        sum_val = New_Die_A[i] + New_Die_B[i]
        expected_count = total_combinations * probability[sum_val - 2]  # Expected count based on probability
        current_count = sum_counts[sum_val - 2]  # Current count of this sum
        if current_count < expected_count:
            # Add spots to Die B to match expected count
            while current_count < expected_count:
                New_Die_B[i] += 1
                current_count += 1

    return New_Die_A, New_Die_B  # Return the transformed dice




# Test the function
Die_A = [1, 2, 3, 4, 5, 6]
Die_B = Die_A
New_Die_A, New_Die_B = undoom_dice(Die_A, Die_B)
print("New Die A:", New_Die_A)
print("New Die B:", New_Die_B)

File: dice.js
// Part A

// 1. Calculate Total Combinations
const totalCombinations = 6 * 6; // Total combinations when rolling two six-sided dice
console.log("Total combinations:", totalCombinations);

// 2. Calculate Distribution of All Possible Combinations
const distribution = [];
for (let i = 1; i <= 6; i++) {
    for (let j = 1; j <= 6; j++) {
        const sum = i + j;
        distribution[sum] = (distribution[sum] || 0) + 1; // Increment count for each sum
    }
}
console.log("Distribution of combinations:", distribution);

// 3. Calculate Probability of All Possible Sums
const probability = distribution.map(count => count / totalCombinations); // Probability = count / total
console.log("Probability of sums:", probability);





// Part B

// Function to Reattach Spots on the Dice while maintaining probabilities
function undoom_dice(Die_A, Die_B) {
    const New_Die_A = [];
    const New_Die_B = [];
    let sumCounts = Array.from({ length: 11 }, () => 0); // Array to store counts of sums 2 to 12

    // Loop through all combinations of Die A and Die B
    for (let i = 0; i < Die_A.length; i++) {
        for (let j = 0; j < Die_B.length; j++) {
            const sum = Die_A[i] + Die_B[j];
            sumCounts[sum - 2]++; // Increment count for this sum (2 to 12)
            if (Die_A[i] <= 4) { // Check condition for Die A
                New_Die_A.push(Die_A[i]); // Keep Die A as is if spots <= 4
                New_Die_B.push(Die_B[j]); // Keep Die B as is
            }
        }
    }

    // Adjust Die B if necessary to maintain probabilities
    for (let i = 0; i < New_Die_B.length; i++) {
        const sum = New_Die_A[i] + New_Die_B[i];
        const expectedCount = totalCombinations * probability[sum]; // Expected count based on probability
        const currentCount = sumCounts[sum - 2]; // Current count of this sum
        if (currentCount < expectedCount) {
            // Add spots to Die B to match expected count
            while (currentCount < expectedCount) {
                New_Die_B[i]++;
                currentCount++;
            }
        }
    }

    return { New_Die_A, New_Die_B }; // Return the transformed dice
}

//We can test the code with the given example inputs


const Die_A = [1, 2, 3, 4, 5, 6];
const Die_B = Die_A;
const { New_Die_A, New_Die_B } = undoom_dice(Die_A, Die_B);
console.log("New Die A:", New_Die_A);
console.log("New Die B:", New_Die_B);
